src/credentials/StaticGroupDataProvider.cpp

```diff
@@ -0,0 +1,780 @@
+        VerifyOrReturnError(static_cast<size_t>(state_index) <= mGroupStateEntriesUsed, CHIP_ERROR_INVALID_ARGUMENT);
+
+        if (static_cast<size_t>(state_index) == mGroupStateEntriesUsed)
+        {
+            // Need to append to end, which impacts indexing
+            appending = true;
+        }
+
+        Fabric * fabric = GetExistingFabricOrAllocateNew(fabric_index);
+        VerifyOrReturnError(fabric, CHIP_ERROR_NO_MEMORY);
+
+        StateEntry * entry = nullptr;
+        uint8_t entry_idx;
+
+        // Case 1: appending a new entry altogether
+        if (appending)
+        {
+            // Search for existing, or unused entry in associated fabric
+            for (entry_idx = 0; entry_idx < kStateEntriesMax; ++entry_idx)
+            {
+                if (!fabric->states[entry_idx].in_use)
+                {
+                    // Found Unused entry
+                    entry = &fabric->states[entry_idx];
+                    break;
+                }
+            }
+        }
+        else
+        {
+            // Case 2: replacing an existing entry's contents
+            GroupStateIndex candidate_mapping = mGroupStateIndices[state_index];
+
+            // Should not happen that our list is inconsistent!
+            VerifyOrReturnError(candidate_mapping.IsValid(), CHIP_ERROR_INTERNAL);
+            entry_idx = candidate_mapping.entry_index;
+
+            Fabric * candidate_fabric = &mFabrics[candidate_mapping.internal_fabric_index];
+            // Avoid overwrite another fabric's entry
+            VerifyOrReturnError(candidate_fabric->in_use && candidate_fabric->fabric_index == fabric_index,
+                                CHIP_ERROR_ACCESS_DENIED);
+
+            entry = &candidate_fabric->states[entry_idx];
+            // Should not happen that our list is inconsistent!
+            VerifyOrReturnError(entry->in_use, CHIP_ERROR_INTERNAL);
+        }
+
+        // Now entry and entry_idx both set properly, or entry is null, so we can continue
+        VerifyOrReturnError(entry, CHIP_ERROR_NO_MEMORY);
+
+        // Set entry contents proper
+        GroupStateListEntry old_state;
+        old_state.group         = entry->group;
+        old_state.key_set_index = entry->key_set_index;
+        old_state.fabric_index  = fabric->fabric_index;
+        old_state.list_index    = state_index;
+
+        entry->group         = state.group;
+        entry->key_set_index = state.key_set_index;
+        entry->in_use        = true;
+
+        // Record global index mapping in the groups state list
+        GroupStateIndex index_mapping;
+        index_mapping.internal_fabric_index = fabric->internal_index;
+        index_mapping.entry_index           = entry_idx;
+        mGroupStateIndices[state_index]     = index_mapping;
+        if (appending)
+        {
+            ++mGroupStateEntriesUsed;
+        }
+
+        if (!appending && mListener)
```

Comment:
Suggest to avoid `!appending` here, and do the following to handle the fact that OnGroupStateChanged has old==nullptr on append.

`mListener->OnGroupStateChanged((appending ? nullptr : &old_state), &new_state);`

Reply:
Applied.